### PR TITLE
Enable Multi-LoRA support for HPU

### PR DESCRIPTION
**Pull request by JHLEE17**
Thank you for your great works. @hlahkar and @SanjuCSudhakaran
I'm a ML engineer from SqueezeBits and I'm interested in contributing to enabling multi-LoRA functionality. It looks like you are already making good progress on this. To ensure that our efforts are aligned and to prevent any redundant work, it would be very helpful if you could share the development schedule for this feature. 

Here are the updates included in this PR:

- Fixed lora_packed_nslice for Multi-LoRA.
- Added and passed LoRA layer test for HPU(test_lora_hpu.py).
- Resolved ~~5~~ ~~4~~ 6 out of 6 matching failures in lora_inference.py.
- Passed test_llama_hpu.py without TP.
(disabled warmup due to long duration.)

I look forward to your response and the opportunity to collaborate.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by vivekgoe** [comment]
@JHLEE17 Thank you for your contribution, we will review your PR and get back soon. We will be happy to collaborate with you on this feature.  
We plan on completing following in next few weeks,
-  Support single adapter LoRA for 1, 2, 4 cards (to be verified using tests/lora/test_llama.py).
-  Support multiple adapter LoRA (to be verified using examples/multilora_inference.py).
-  Analyze performance/throughput with LoRA and optimize it (if required).
-  Explore mechanism to test server scenario with LoRA and fix any issues we face with this.
-  Get some of the LoRA related tests here "tests/lora" working on Gaudi/HPU. 

If you could share some more information about real use-cases for LoRA feature being used by SqueezeBit or any examples available outside vllm repository for LoRA feature we can try to figure out what is needed to support those on Gaudi/HPU.

**Comment by JHLEE17** [comment]
@vivekgoe We are primarily interested in ensuring that all test cases pass first. As such, our current focus is on supporting TP on multi-HPU and aligning the multi-LoRA functionality with GPU. It looks like you are handling TP very well, especially in the habana_main branch. Therefore, I would like to concentrate on aligning the multi-LoRA functionality first.

Could you please share your thoughts on this through the PR review? I appreciate your collaboration and look forward to your feedback.

File: examples/lora_inference.py [hunk 1]
```diff
@@ -38,7 +38,7 @@
 for i, output in enumerate(outputs):
     prompt = output.prompt
     generated_text = output.outputs[0].text
-    matching = expected_output[i].lower() == generated_text.lower()
+    matching = expected_output[i] == generated_text
```
[the remark this answers is not in the document]
Comment (SanjuCSudhakaran) [reply examples/lora_inference.py:41]: Yes, you are right. We have already removed this conversion from the test in our latest push.

File: tests/lora/utils.py [hunk 2]
```diff
@@ -28,16 +28,16 @@ def init_random_lora(self,
             lora_alpha=1,
             lora_a=torch.rand([weight.shape[1], rank],
                               dtype=weight.dtype,
-                              device="cuda"),
+                              device="hpu"),
```
Comment (hlahkar) [inline tests/lora/utils.py:31]: We need to protect this in a way; it work with cuda also. You can refer to vllm/lora/model.py:96; in this branch

File: vllm/worker/habana_worker.py [hunk 3]
```diff
@@ -162,7 +162,7 @@ def initialize_cache(self, num_gpu_blocks: int,
         self.cache_config.num_cpu_blocks = num_cpu_blocks
 
         self._init_cache_engine()
-        self._warm_up_model()
+        # self._warm_up_model()
```
Comment (hlahkar) [inline vllm/worker/habana_worker.py:165]: No need to comment this; we can skip the warmup by setting the below environment variable 

VLLM_SKIP_WARMUP=true

File: tests/lora/test_llama_hpu.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 import vllm
 from vllm.lora.request import LoRARequest
 
-from conftest import cleanup
+from .conftest import cleanup
```
Comment (hlahkar) [inline tests/lora/test_llama_hpu.py:9]: we have already created a test_llama_hpu.py to replicate these tests

File: tests/lora/test_lora_hpu.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    module_name = "module"
+    weight = torch.rand([m, n], device="hpu", dtype=dtype)
+
+    manager.init_random_lora(module_name, weight, rank=rank)
+    lora = manager.get_module_lora(module_name)
+
+    input = torch.rand(k, n, device="hpu", dtype=dtype)
+    expected = input @ lora.lora_a @ lora.lora_b * lora.scaling
+
+    lora_a_stack = torch.zeros(8,
```
Comment (SanjuCSudhakaran) [inline tests/lora/test_lora_hpu.py:41]: We handle no LoRA case using zero valued A and B tensors. So need to append these matrices at the end. So the size will be max_loras + 1. Also please handle the for loop (line number 53) and indices ((line number 60) in the test case such that the last index of the LoRA matrices are untouched.

File: examples/lora_inference.py [hunk 6]
```diff
@@ -4,7 +4,7 @@
 
 sql_lora_path = snapshot_download(repo_id="yard1/llama-2-7b-sql-lora-test")
 
-llm = LLM(model="meta-llama/Llama-2-7b-hf", enable_lora=True, max_num_seqs=16, dtype='bfloat16')
+llm = LLM(model="meta-llama/Llama-2-7b-hf", enable_lora=True, max_num_seqs=16)
```
[the remark this answers is not in the document]
Comment (vivekgoe) [reply examples/lora_inference.py:7]: @JHLEE17 Although Gaudi2 supports FP16, we use BF16 for most of our development and testing. So we would prefer to concentrate only on BF16 (unless there is a good reason for using FP16). 
Regarding Gaudi-BF16 != A100-BF16 we will check if its a real issue or not.

**Comment by vivekgoe** [comment]
@JHLEE17  please rebase to latest on this branch, that should address some of your comments. Please also check comments from @SanjuCSudhakaran and @hlahkar . Thanks.

File: examples/lora_inference.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (JHLEE17) [reply examples/lora_inference.py:7]: @vivekgoe While the generated tokens of the Llama-2-7B model (FP16) were identical on both Gaudi and A100 devices, there were slight differences in the logit values. Additionally, for the Llama-3-8B model, each case (whether FP16/BF16 on Gaudi/A100) produced slightly different results. I suspect these inconsistency are inherent to floating point operations in generative models. I would greatly appreciate any insights or opinions you could provide on this matter.

File: tests/lora/test_lora_hpu.py (hunk 5 above)
Comment (JHLEE17) [reply tests/lora/test_lora_hpu.py:41]: I completely agree with your suggestion. However, while reviewing the original vllm repo for this task, I noticed that the test_lora.py file has been recently removed in [7ecee34](https://github.com/vllm-project/vllm/commit/7ecee3432110bae563c8756a66b54e5f08dc777d) and seems to have been replaced by other test files. Considering the upstream changes, what do you think about using the alternative test files instead?

**Comment by JHLEE17** [comment]
> @JHLEE17 please rebase to latest on this branch, that should address some of your comments. Please also check comments from @SanjuCSudhakaran and @hlahkar . Thanks.

@vivekgoe  Thank you for your comments and valuable feedback. I have reviewed all the suggestions provided. Some of the modifications in this PR were made specifically to replicate my experimental results, so it might be good to ignore these changes into the main branch.

This PR was based on commit [301579d](https://github.com/HabanaAI/vllm-fork/commit/1ee15b4aff73e9971178906da0ae9b31b519dd09), and I noticed that most of the changes for multi-LoRA support have been included in the subsequent commit [1ee15b4](https://github.com/HabanaAI/vllm-fork/commit/1ee15b4aff73e9971178906da0ae9b31b519dd09). Therefore, rebasing might not result in significant differences.

Considering this, do you think it would be more appropriate to close this PR? Your guidance on this matter would be good.

File: examples/lora_inference.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vivekgoe) [reply examples/lora_inference.py:7]: @JHLEE17 Slight differences in logit values between Gaudi/A100 are expected due to following reasons,
-  For operations on large tensor inputs Gaudi & A100 may slice the input tensors and combine back outputs differently, mathematically both produce same output but due to quantization errors inherent to floating point representation this will result in small differences in final outputs.  However if we use some utility like lm_harness to evaluate quality of output generated over a input dataset with multiple prompts then we do expect to see same accuracy results for Gaudi/A100. 
-  For some operations which may use random number generators (e.g. sampling), both Gaudi & A100 will choose from same distributions but actual values picked (even with same seed) on a given iteration will be different because random number generator implementations will be different.

File: tests/lora/test_lora_hpu.py (hunk 5 above)
Comment (vivekgoe) [reply tests/lora/test_lora_hpu.py:41]: @JHLEE17 For now we are planning to add copies of original test with a "_hpu" suffix into HabanaAI/vllm-fork, therefore it should be ok to take this test into vllm-fork. Later when we upstream these changes to vllm-project we will figure out correct approach (maybe create a separate folder for hpu tests).

**Comment by vivekgoe** [comment]
@JHLEE17 I agree that since most of the changes suggested in this PR have been included in subsequent commits we can close this PR. However I would still want to take test_lora_hpu.py changes (aligned to latest commit on enable_lora branch).  Please let us know if it is ok for us to do this or if you would like to do it yourself as part of this PR.

**Comment by JHLEE17** [comment]
@vivekgoe As your suggestion, I have rebased to the latest commit and completed the work on test_lora_hpu.py. It would be great if you could confirm it.
